File: src/dotnet/Ansight.Tools.FileSystem/CopyFileTool.cs
namespace Ansight.Tools.FileSystem;

public sealed class CopyFileTool : ITool
{
    private readonly FileSystemToolsOptions options;

    public CopyFileTool(FileSystemToolsOptions? options = null)
    {
        this.options = options ?? FileSystemToolsOptions.Default;
    }

    public string Category => "files";

    public ToolPolicy Policy => ToolPolicy.Write;

    public string Id => FileSystemToolIds.CopyFile;

    public string Name => "Copy File";

    public string Description => "Copies a file between sandbox-constrained paths.";

    public string Keywords => "filesystem file copy duplicate sandbox";

    public ToolSchema ArgumentsSchema => FileSystemToolSchemas.CopyFileArguments;

    public ToolSchema ResultSchema => FileSystemToolSchemas.CopyFileResult;

    public Task<ToolResult> Execute(IReadOnlyDictionary<string, string> arguments)
    {
        ArgumentNullException.ThrowIfNull(arguments);

        try
        {
            var roots = FileSystemSandbox.GetRoots(options);
            var sourceFile = FileSystemWriteHelpers.ResolveFile(
                arguments,
                roots,
                pathKey: "sourcePath",
                rootKey: "root",
                requireExisting: true);
            var destinationFile = FileSystemWriteHelpers.ResolveDestinationFile(arguments, roots, sourceFile);
            FileSystemWriteHelpers.EnsureDifferentPaths(sourceFile, destinationFile);

            var overwrite = FileSystemSandbox.GetBoolean(arguments, "overwrite", defaultValue: false);
            var createDirectory = FileSystemSandbox.GetBoolean(arguments, "createDirectory", defaultValue: true);
            var overwritten = FileSystemWriteHelpers.EnsureDestinationFileCanBeWritten(destinationFile, overwrite);
            var createdDirectory = FileSystemWriteHelpers.EnsureParentDirectoryExists(destinationFile, createDirectory);

            File.Copy(sourceFile.FullPath, destinationFile.FullPath, overwrite);

            var payload = FileSystemWriteHelpers.CreateFileTransferPayload(
                "copied",
                sourceFile,
                destinationFile,
                roots,
                overwritten,
                createdDirectory);

            return Task.FromResult(ToolResult.Success(payload));
        }
        catch (Exception exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "filesystem_copy_failed"));
        }
    }
}

File: src/ios/Sources/AnsightToolsFileSystem/CopyFileTool.swift
import AnsightCore
import Foundation

public final class CopyFileTool: AnsightTool {
    private let options: AnsightFileSystemToolsOptions

    public init(options: AnsightFileSystemToolsOptions = .default) {
        self.options = options
    }

    public var descriptor: AnsightToolDescriptor {
        AnsightToolDescriptor(
            id: AnsightFileSystemToolIds.copyFile,
            name: "Copy File",
            description: "Copies a sandboxed file to another approved sandbox path.",
            category: "files",
            policy: .write,
            keywords: "filesystem file copy sandbox",
            argumentsSchema: AnsightFileSystemToolSchemas.copyFileArguments,
            resultSchema: AnsightFileSystemToolSchemas.copyFileResult
        )
    }

    public func execute(arguments: [String: String]) throws -> AnsightToolExecutionResult {
        do {
            let roots = try AnsightFileSystemSandbox.roots(options: options)
            let source = try AnsightFileSystemSandbox.resolvePath(
                arguments: arguments,
                roots: roots,
                pathKey: "sourcePath",
                requireExisting: true,
                expectDirectory: false
            )
            let destination = try AnsightFileSystemSandbox.resolveDestinationPath(
                arguments: arguments,
                roots: roots,
                sourceRoot: AnsightFileSystemSandbox.root(for: source)
            )
            let overwrite = try AnsightFileSystemSandbox.boolean(arguments, key: "overwrite", defaultValue: false)
            let createDirectory = try AnsightFileSystemSandbox.boolean(arguments, key: "createDirectory", defaultValue: false)
            let overwritten = FileManager.default.fileExists(atPath: destination.fullPath)
            let createdDirectory = try AnsightFileSystemSandbox.ensureWritableDestination(
                destination,
                overwrite: overwrite,
                createDirectory: createDirectory
            )

            if overwritten {
                try FileManager.default.removeItem(atPath: destination.fullPath)
            }

            try FileManager.default.copyItem(atPath: source.fullPath, toPath: destination.fullPath)
            return try transferResult(
                operation: "copied",
                source: source,
                destination: destination,
                roots: roots,
                overwritten: overwritten,
                createdDirectory: createdDirectory
            )
        } catch {
            return .failure(error.localizedDescription, errorCode: "filesystem_copy_failed")
        }
    }

    private func transferResult(
        operation: String,
        source: AnsightFileSystemResolvedPath,
        destination: AnsightFileSystemResolvedPath,
        roots: [AnsightFileSystemRoot],
        overwritten: Bool,
        createdDirectory: Bool
    ) throws -> AnsightToolExecutionResult {
        let attributes = try AnsightFileSystemIO.attributes(path: destination.fullPath)
        guard case .object(var payload) = AnsightFileSystemContentDescriptor.resolvedFilePayload(
            resolvedFile: destination,
            roots: roots,
            attributes: attributes
        ) else {
            return .failure("Failed to build copy payload.", errorCode: "filesystem_copy_failed")
        }

        payload["operation"] = .string(operation)
        payload["sourceRootAlias"] = .string(source.rootAlias)
        payload["sourceRootPath"] = .string(source.rootPath)
        payload["sourceFilePath"] = .string(source.fullPath)
        payload["sourceRelativePath"] = .string(AnsightFileSystemSandbox.relativePath(rootPath: source.rootPath, fullPath: source.fullPath))
        payload["destinationRootAlias"] = .string(destination.rootAlias)
        payload["destinationRootPath"] = .string(destination.rootPath)
        payload["destinationFilePath"] = .string(destination.fullPath)
        payload["destinationRelativePath"] = .string(AnsightFileSystemSandbox.relativePath(rootPath: destination.rootPath, fullPath: destination.fullPath))
        payload["overwritten"] = .bool(overwritten)
        payload["createdDirectory"] = .bool(createdDirectory)
        payload["capturedAtUtc"] = .string(AnsightClock.isoNow())
        return .success(.object(payload))
    }
}
